torch/csrc/jit/codegen/cuda/lower_predicate.cpp

```diff
@@ -159,8 +230,66 @@ class PredicateAnalyzer : public OptOutDispatch {
     // larger than the axis size, it's not safe to skip predication
 
     // Check that parallel dimension will not generate out of bound index
-    if (!(producer->getMemoryType() == MemoryType::Local &&
-          consumer->getMemoryType() == MemoryType::Local)) {
+    if (producer->getMemoryType() == MemoryType::Global ||
+        consumer->getMemoryType() == MemoryType::Global) {
+      return true;
+    }
+
+    // This is initial step to gradually remove predicates around
+    //  sharedmem access in suitable situations.
+    // Using an additional variable to track the predicate-on reasons
+    //  when the predicate around shared mem cannot be removed.
+    bool needs_sharedmem_addr_pred = false;
+    if (producer->getMemoryType() == MemoryType::Shared ||
+        consumer->getMemoryType() == MemoryType::Shared) {
+      // Indexing is based on consumer leaf ids so check the consumer.
```

Comment:
IIUC, we don't need to check the producer here as it's going to be analyzed in `needsPrediate(IterDomain*)`. Am I right?

Reply:
I think we check consumer only here because we read producer with consumer schedule when producer is in shared mem.